### PR TITLE
[7.x] Fix `whereNull`/`whereNotNull` for json in MySQL

### LINKED ISSUE #31473: JSON + whereNull/whereNotNull is non-intuitive or broken
- Laravel Version: 5.8 / 6.x
- PHP Version: 7.2.0
- Database Driver & Version: MySQL 5.7 (not sure about other dbs)

### Description:

The `whereNull('a->b')` will generate ``json_unquote(json_extract(`a`, '$."b"')) is null``  that doesn't work as expected:

```sql
set @a='{"a":3,"b":null}';

select 
json_unquote(json_extract(@a, '$."b"')) IS NULL, -- = false (but value is null => should be true)
json_unquote(json_extract(@a, '$."c"')) IS NULL; -- = true
```

image

### Desired behavior:
1) IMHO `whereNull` should cover both cases: 
    1) key doesn't exist (fine now)
    2) key exists but the value is `null` (the `JSON_TYPE(json_extract(@a, '$."b"')) = 'NULL'` can be used)
2) (or) will be probably good to have a notice in docs about this non-intuitive `whereNull` behavior.

Also, this issue similar to #27696

### PR DESCRIPTION
**Pull request by LastDragon-ru**
This PR fixes #31473, please see issue for more details, in short:

```sql
set @a='{"value":null}';

select
-- Actual behavior whereNull/whereNotNull
json_unquote(json_extract(@a, '$."value"')) is null,     -- false but should be true because value = null
json_unquote(json_extract(@a, '$."value"')) is not null, -- true  but should be false 

-- Expected
(json_extract(@a, '$."value"') is null     OR  json_type(json_extract(@a, '$."value"')) = 'NULL'),  -- true, ok
(json_extract(@a, '$."value"') is not null AND json_type(json_extract(@a, '$."value"')) != 'NULL')  -- false, ok
;
```

_PS: I tested it only for MySQL, probably would be good also test for MariaDB and Percona._

### CODE REVIEW

**Comment by if4lcon** [comment]
`whereJsonLength` works for me

```
//where not null
->whereJsonLength('json', '>=', 1); 
```

```
//where  null
->whereJsonLength('json', '<=', 0); 
```

**Comment by LastDragon-ru** [comment]
> whereJsonLength works for me

@if4lcon is it work if key does not exist?

**Comment by taylorotwell** [comment]
@staudenmeir do you have any feedback on this?

**Comment by staudenmeir** [comment]
Looks good to me. Great tests!

File: tests/Integration/Database/DatabaseMySqlConnectionTest.php [hunk 1]
```diff
@@ -71,4 +71,39 @@ public function testFloatValueStoredCorrectly(): void
...
+        return [
+            //                                  exists?    key                  json
+            'key not exists'                 => [true,     'invalid'],
+            'key exists and null'            => [true,     'value',             ['value' => null]],
+            'key exists and "null"'          => [false,    'value',             ['value' => 'null']],
+            'key exists and not null'        => [false,    'value',             ['value' => false]],
+            'nested key not exists'          => [true,     'nested->invalid'],
+            'nested key exists and null'     => [true,     'nested->value',     ['nested' => ['value' => null]]],
+            'nested key exists and "null"'   => [false,    'nested->value',     ['nested' => ['value' => 'null']]],
+            'nested key exists and not null' => [false,    'nested->value',     ['nested' => ['value' => false]]],
```
Comment (driesvints) [inline tests/Integration/Database/DatabaseMySqlConnectionTest.php:106]: Please remove the indentation here.

**Review by driesvints — DISMISSED** [review]

File: src/Illuminate/Database/Query/Grammars/MySqlGrammar.php [hunk 2]
```diff
@@ -244,4 +244,35 @@ protected function wrapJsonBooleanSelector($value)
 
         return 'json_extract('.$field.$path.')';
     }
+
+    protected function whereNull(Builder $query, $where)
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Grammars/MySqlGrammar.php:248]: Missing phpdoc.

File: src/Illuminate/Database/Query/Grammars/MySqlGrammar.php [hunk 3]
```diff
@@ -244,4 +244,35 @@ protected function wrapJsonBooleanSelector($value)
...
+
+            [$field, $path] = $this->wrapJsonFieldAndPath($where['column']);
+
+            return '(json_extract('.$field.$path.') is null OR json_type(json_extract('.$field.$path.')) = \'NULL\')';
+        }
+
+        return parent::whereNull($query, $where);
+    }
+
+    protected function whereNotNull(Builder $query, $where)
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Grammars/MySqlGrammar.php:268]: Same

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by taylorotwell** [comment]
Fixed PHP doc issues.

**Pull request merged**